### PR TITLE
2229: Update signature for truncate function

### PR DESCRIPTION
**Pull request by arora-aman**
r? @nikomatsakis

### CODE REVIEW

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 1]
```diff
@@ -1841,31 +1837,28 @@ impl<'a, 'tcx> euv::Delegate<'tcx> for InferBorrowKind<'a, 'tcx> {
...
+    mut curr_mode: ty::UpvarCapture<'tcx>,
 ) -> (Place<'tcx>, ty::UpvarCapture<'tcx>) {
-    if place.projections.is_empty() {
-        // Nothing to do here
-        return (place, curr_mode);
-    }
-
     if place.base_ty.is_unsafe_ptr() {
-        return truncate_place_to_len(place, curr_mode, 0);
+        truncate_place_to_len_and_update_capture_kind(&mut place, &mut curr_mode, 0);
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/upvar.rs:1844]: we used to return here, but now we don't? maybe this function also wants to take `&mut` arguments?

File: compiler/rustc_typeck/src/check/upvar.rs [hunk 2]
```diff
@@ -1880,7 +1873,7 @@ fn restrict_capture_precision<'tcx>(
     place: Place<'tcx>,
     curr_mode: ty::UpvarCapture<'tcx>,
 ) -> (Place<'tcx>, ty::UpvarCapture<'tcx>) {
-    let (place, curr_mode) = restrict_precision_for_unsafe(place, curr_mode);
+    let (mut place, mut curr_mode) = restrict_precision_for_unsafe(place, curr_mode);
```
Comment (nikomatsakis) [inline compiler/rustc_typeck/src/check/upvar.rs:1876]: as above, I suspect this could take `&mut` place arguments

**Review by nikomatsakis — CHANGES_REQUESTED** [review]
I think we should push the `&mut` arguments through more thoroughly

File: compiler/rustc_typeck/src/check/upvar.rs (hunk 1 above)
Comment (arora-aman) [reply compiler/rustc_typeck/src/check/upvar.rs:1844]: we just return once at the bottom. Otherwise each of these has a `return (place, curr_mode)`

**Comment by nikomatsakis** [comment]
What do you think about having the general pattern be to take `&mut` and not to return a tuple, @arora-aman ?

**Comment by arora-aman** [comment]
> What do you think about having the general pattern be to take `&mut` and not to return a tuple, @arora-aman ?

I personally feel returning is more functional  and gives the reader of the code an exact idea of what's going on / being upadted (though so does the name). 

Also returning the tuple makes it closer to how the algorithm is written out in the reference. But If you prefer `&mut` everywhere I don't mind making the change :)

**Comment by nikomatsakis** [comment]
@arora-aman I see your point. Let's keep it as is.

@bors  r+

**Pull request merged**